### DISCUSSION TITLE
PersistAll extension for replying

### DISCUSSION THREAD

**Discussion by Haukinger** [akkadotnet/akka.net#7407] [Ideas]
Most of my persistent actors use `PersistAll` with some kind of flag to reply to the sender, i.e.

            PersistAll( events,
                        @event =>
                        {
                            _state = _state.Process( @event );

                            if (!sentReply)
                            {
                                sender.Tell( reply );
                                sentReply = true;
                            }
                            if (LastSequenceNr % 10 == 0)
                                SaveSnapshot( _state );
                        } );

To me, it makes sense to have a callback that's called after processing the events into the state, like

            PersistAll( events,
                        @event =>
                        {
                            _state = _state.Process( @event );

                            if (LastSequenceNr % 10 == 0)
                                SaveSnapshot( _state );
                        },
                        () => 
                        {
                            sender.Tell( reply );
                        } );

or perhaps

            PersistAll( events,
                        @event =>
                        {
                            _state = _state.Process( @event );

                            if (LastSequenceNr % 10 == 0)
                                SaveSnapshot( _state );
                        } )
               .ContinueWith( _ => sender.Tell( reply ) );

This would be nice with edge cases where occasionally no events are generated but you still want to send a reply.

**Comment by Aaronontheweb** [comment]
You know, I've thought of this myself - the boilerplate I have to do around sending that reply at the very end is kind of a nuisance. I wonder if it'd be better if we changed the signature so you get the array of all events back and you can just do your stuff that way? Or would that just introduce new problems because now you have to loop through it all yourself?

**Reply by aloker** [reply]
That should be ok, too. In the simplest case, that part of the typical pattern would boil down to

```
PersistAll(events).ForEach(Apply);
Sender.Tell(response);
```

Not bad at all IMO. And if the per-event logic is more complex, a `foreach(var @event in PersistAll(events)){}` still isn't too bad.
